### PR TITLE
Provider vsphere finalize method

### PR DESCRIPTION
**Pull request by jacobweinstock**
This PR implements the finalize method of downloading files from an engine's deliverables endpoint and putting them in the `.cake/<cluster-name>/` directory.

### CODE REVIEW

File: docs/Provider-Finalize.md [hunk 1]
```diff
@@ -0,0 +1,24 @@
+# Provider Interface
```
Comment (bgentry-ntap) [inline docs/Provider-Finalize.md:1]: The filename should probably be Provider or Bootstrapper-Interface since this is more than just the finalize

File: docs/Provider-Finalize.md [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+The provider package supplies a bootstrapper interface to handle the creation of infrastructure, running cake engines, monitoring progress and downloading deployment files.
+
+This doc will outline the methods of the bootstrapper interface in a little more detail.
+
+### Methods
+
+- #### Client
```
Comment (bgentry-ntap) [inline docs/Provider-Finalize.md:9]: I wonder if we can just generate go docs and reference the docs in code? Probably not for this PR but something to consider.

File: pkg/engine/rkecli/rkecli.go [hunk 3]
```diff
@@ -90,6 +90,10 @@ func (c *MgmtCluster) InstallControlPlane() error {
 
 // Spec returns the Spec
 func (c *MgmtCluster) Spec() engine.MgmtCluster {
+	c.MgmtCluster.FileDeliverables = []string{
+		c.RKEConfigPath,
+		"/kube_config_rke-config.yml",
```
Comment (bgentry-ntap) [inline pkg/engine/rkecli/rkecli.go:95]: Thoughts on just making this kube_config or rke_kube_config? Mix of snake and spinal case is weird and we don't have to take the weirdness from rke to our endpoints.

File: pkg/progress/progress.go [hunk 4]
```diff
@@ -31,25 +44,41 @@ func init() {
...
+			dv = append(dv, DeliverableInfo{
+				Url:     uri,
+				FileExt: filepath.Ext(fileDeliverables[x]),
+			})
 		}
-		fmt.Fprintf(w, string(kconfig))
+	}
+	http.HandleFunc(URIDeliverable, func(w http.ResponseWriter, r *http.Request) {
+		json.NewEncoder(w).Encode(dv)
 	})
```
Comment (bgentry-ntap) [inline pkg/progress/progress.go:82]: Thoughts on logging all the endpoints?

File: pkg/progress/progress.go [hunk 5]
```diff
@@ -31,25 +44,41 @@ func init() {
...
 		logs, _ := ioutil.ReadFile(logfile)
 		fmt.Fprintf(w, string(logs))
 	})
-	http.HandleFunc("/kubeconfig", func(w http.ResponseWriter, r *http.Request) {
-		kconfig, _ := ioutil.ReadFile(kubeconfig)
-		if len(kconfig) == 0 {
-			w.WriteHeader(http.StatusInternalServerError)
+
+	var dv []DeliverableInfo
+	for x := 0; x < len(fileDeliverables); x++ {
```
Comment (bgentry-ntap) [inline pkg/progress/progress.go:64]: range would be a little cleaner here

File: pkg/progress/progress.go [hunk 6]
```diff
@@ -59,3 +88,20 @@ func ServeDuration() {
 		time.Sleep(1 * time.Hour)
 	}
 }
+
+func DownloadTxtFile(url string, downloadLocation string) error {
+	resp, err := http.Get(url)
+	if err != nil {
+		return err
```
Comment (bgentry-ntap) [inline pkg/progress/progress.go:95]: adding context to all the error returns would be good. applies throughout

File: pkg/progress/progress_test.go [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+	"testing"
+)
+
+func TestMain(m *testing.M) {
```
Comment (bgentry-ntap) [inline pkg/progress/progress_test.go:13]: Thoughts on integration test folder/structure? Unit tests should not make OS calls or serve full web stack

File: pkg/progress/progress.go (hunk 6 above)
Comment (jacobweinstock) [reply pkg/progress/progress.go:95]: will update

File: pkg/progress/progress.go (hunk 5 above)
Comment (jacobweinstock) [reply pkg/progress/progress.go:64]: will update

File: pkg/progress/progress.go (hunk 4 above)
Comment (jacobweinstock) [reply pkg/progress/progress.go:82]: hmmm....i could be swayed either way. i see cases for both.

File: pkg/engine/rkecli/rkecli.go (hunk 3 above)
Comment (jacobweinstock) [reply pkg/engine/rkecli/rkecli.go:95]: yeah, i like either one of those. Can the rkecli engine change the name of that file? `FileDeliverables` is not setup to change the name of any files. i guess it could but i think i'd prefer the engine determine the name.

File: docs/Provider-Finalize.md (hunk 1 above)
Comment (jacobweinstock) [reply docs/Provider-Finalize.md:1]: will update

File: docs/Provider-Finalize.md (hunk 2 above)
Comment (jacobweinstock) [reply docs/Provider-Finalize.md:9]: i like this idea. i can just move this info to comments on the methods for now.

File: pkg/engine/rkecli/rkecli.go (hunk 3 above)
Comment (bgentry-ntap) [reply pkg/engine/rkecli/rkecli.go:95]: Yeah I'm good with the engine doing that 👍

File: pkg/progress/progress.go (hunk 4 above)
Comment (bgentry-ntap) [reply pkg/progress/progress.go:82]: I'm just lazy and like clicking links from the console. Might also help users who inevitably won't read the docs 😁

Comment (jacobweinstock) [reply pkg/progress/progress.go:82]: sure thing. i'll add some logging to show what endpoints were served

File: pkg/progress/progress_test.go (hunk 7 above)
Comment (jacobweinstock) [reply pkg/progress/progress_test.go]: we should talk about a long term solution. for the moment i renamed this file with `_integration_test` and add a build flag so it wont be run with `make test`

File: pkg/progress/progress.go (hunk 4 above)
Comment (jacobweinstock) [reply pkg/progress/progress.go:93]: output from successful run 
```bash
./bin/cake-darwin deploy -d rke -f examples/config-rke.yaml
INFO[0000] generated a cluster name:  wondrous-tarpon
INFO[0000] creating config directory: /Users/weinstoj/.cake/wondrous-tarpon
INFO[0000] Welcome to Mission Control
INFO[0000] Let's launch a cluster                        ClusterName=rke-mgmt-cluster ControlPlaneMachineCount=1 workerMachineCount=2
INFO[0000] progress event                                fields.level=info fields.msg="Connecting to provider" type=progress
INFO[0001] progress event                                fields.level=info fields.msg="Preparing environment" type=progress
INFO[0028] progress event                                fields.level=info fields.msg="Provisioning cluster" type=progress
INFO[0084] progress event                                fields.level=info fields.msg="IP received for rke-mgmt-cluster-controlplane-1: 172.60.0.32" type=progress
INFO[0085] progress event                                fields.level=info fields.msg="IP received for rke-mgmt-cluster-worker-1: 172.60.0.33" type=progress
INFO[0086] progress event                                fields.level=info fields.msg="IP received for rke-mgmt-cluster-worker-2: 172.60.0.126" type=progress
INFO[0119] progress event                                fields.level=info fields.msg="Provision Progress" type=progress
INFO[0244] progress event                                fields.level=info fields.msg="type: progress, msg: serving file: /tmp/cake.log at http://172.60.0.32:8081/log, level: debug" type=progress
INFO[0244] progress event                                fields.level=info fields.msg="type: progress, msg: serving file: /rke-config.yml at http://172.60.0.32:8081/deliverable/rke-config, level: debug" type=progress
INFO[0244] progress event                                fields.level=info fields.msg="type: progress, msg: serving file: /kube_config_rke-config.yml at http://172.60.0.32:8081/deliverable/kube_config_rke-config, level: debug" type=progress
INFO[0246] progress event                                fields.level=info fields.msg="type: progress, msg: TODO: provide required commands, level: " type=progress
INFO[0246] progress event                                fields.level=info fields.msg="type: progress, msg: CreateBootstrap nothing to do..., level: " type=progress
INFO[0246] progress event                                fields.level=info fields.msg="type: progress, msg: InstallControlPlan nothing to do..., level: " type=progress
INFO[0246] progress event                                fields.level=info fields.msg="type: progress, msg: install HA rke cluster, level: " type=progress
INFO[0618] progress event                                fields.level=info fields.msg="type: progress, msg: added rancher-stable helm chart, level: " type=progress
INFO[0619] progress event                                fields.level=info fields.msg="type: progress, msg: added cert-manager helm chart, level: " type=progress
INFO[0619] progress event                                fields.level=info fields.msg="type: progress, msg: created namespaces, level: " type=progress
INFO[0627] progress event                                fields.level=info fields.msg="type: progress, msg: installed cert-manager CRD, level: " type=progress
INFO[0627] progress event                                fields.level=info fields.msg="type: progress, msg: updated helm chart, level: " type=progress
INFO[0635] progress event                                fields.level=info fields.msg="type: progress, msg: helm installed cert-manager, level: " type=progress
INFO[0635] progress event                                fields.level=info fields.msg="type: progress, msg: waiting for cert-manager to be ready, level: " type=progress
INFO[0711] progress event                                fields.level=info fields.msg="type: progress, msg: helm installing rancher, level: " type=progress
INFO[0711] progress event                                fields.level=info fields.msg="type: progress, msg: suppressing error running rancher helm install: err: err: exit status 1, stderr: Error: Internal error occurred: failed calling webhook \"webhook.cert-manager.io\": Post https://cert-manager-webhook.cert-manager.svc:443/mutate?timeout=30s: dial tcp 10.43.200.79:443: connect: connection refused\n, cmd: helm install rancher rancher-stable/rancher --version=2.4.3 --namespace=cattle-system --kubeconfig=/kube_config_rke-config.yml --set hostname=my.rancher.org,certmanager.version=v0.15.0, stderr: Error: Internal error occurred: failed calling webhook \"webhook.cert-manager.io\": Post https://cert-manager-webhook.cert-manager.svc:443/mutate?timeout=30s: dial tcp 10.43.200.79:443: connect: connection refused\n, cmd: helm install rancher rancher-stable/rancher --version=2.4.3 --namespace=cattle-system --kubeconfig=/kube_config_rke-config.yml --set hostname=my.rancher.org,certmanager.version=v0.15.0, level: " type=progress
INFO[0711] progress event                                fields.level=info fields.msg="type: progress, msg: waiting for rancher to be ready, level: " type=progress
INFO[0763] progress event                                fields.level=info fields.msg="type: progress, msg: waiting for nginx ingress to be ready, level: " type=progress
INFO[0763] progress event                                fields.level=info fields.msg="type: progress, msg: rancher Issuer failed to deploy, recreating: err: err: exit status 1, stderr: Error from server (NotFound): issuers.cert-manager.io \"rancher\" not found\n, cmd: kubectl wait issuer rancher --for condition=ready --namespace=cattle-system --kubeconfig=/kube_config_rke-config.yml, stderr: Error from server (NotFound): issuers.cert-manager.io \"rancher\" not found\n, cmd: kubectl wait issuer rancher --for condition=ready --namespace=cattle-system --kubeconfig=/kube_config_rke-config.yml, level: " type=progress
INFO[0764] progress event                                fields.level=info fields.msg="type: progress, msg: Make sure hostname my.rancher.org resolves to 172.60.0.126 or a worker node IP, level: " type=progress
INFO[0764] progress event                                fields.level=info fields.msg="type: progress, msg: HA rancher install complete: https://my.rancher.org, level: " type=progress
INFO[0764] progress event                                fields.level=info fields.msg="type: progress, msg: TODO: install addons, level: " type=progress
INFO[0764] progress event                                fields.level=info fields.msg=Finalizing type=progress
INFO[0765] progress event                                fields.level=info fields.msg="all files from the cluster deployment can be found here: /Users/weinstoj/.cake/wondrous-tarpon/" type=progress
```

File: Makefile [hunk 8]
```diff
@@ -62,6 +62,11 @@ test: ## Test with go test
 	@echo "$$STATIK_FILE" > pkg/util/statik/statik.go
 	go test -v -covermode=count ./...
 
+.PHONY: integration
+integration: ## Integration tests with go test
+	@echo "$$STATIK_FILE" > pkg/util/statik/statik.go
+	go test -tags=integration -v -covermode=count ./...
```
Comment (bgentry-ntap) [inline Makefile:68]: Nice!!

File: pkg/engine/interface.go [hunk 9]
```diff
@@ -93,3 +95,13 @@ func Run(c Cluster) error {
 
 	return nil
 }
+
+func getLocalIP() string {
```
Comment (bgentry-ntap) [inline pkg/engine/interface.go:99]: slick

File: pkg/progress/progress.go [hunk 10]
```diff
@@ -44,7 +44,8 @@ func init() {
 	responseBody.Messages = []string{}
 }
 
-func Serve(logfile string, port string, status Events, fileDeliverables []string) {
+func Serve(logfile string, ip, port string, status Events, fileDeliverables []string) {
+	fullURL := fmt.Sprintf("http://%s:%s", ip, port) + "%s"
```
Comment (bgentry-ntap) [inline pkg/progress/progress.go:48]: 😬 URL struct might be good here

File: pkg/provider/interface.go [hunk 11]
```diff
@@ -54,6 +54,7 @@ func Run(b Bootstrapper) error {
 		Msg:   "Preparing environment",
 		Level: "info",
 	})
+	defer b.Finalize()
```
Comment (bgentry-ntap) [inline pkg/provider/interface.go:57]: Just curious, why this change?

Comment (jacobweinstock) [reply pkg/provider/interface.go:57]: yeah, sorry. this is a throw in at the end here. i can remove it. the idea here that if progress or provider methods fail we still download the logs.

File: pkg/progress/progress.go (hunk 10 above)
Comment (jacobweinstock) [reply pkg/progress/progress.go:48]: yeah, will update

File: pkg/provider/interface.go (hunk 11 above)
Comment (bgentry-ntap) [reply pkg/provider/interface.go:57]: Oooh I like that. I say leave it. Good catch

File: pkg/progress/progress.go (hunk 10 above)
Comment (jacobweinstock) [reply pkg/progress/progress.go]: updated
